src/celeritas/setup/Problem.cc

```diff
@@ -279,62 +281,96 @@ auto build_along_step(inp::Field const& var_field,
+    inp.detector_labels = optical.detector_labels;
+
+    CELER_ENSURE(inp);
+
+    return std::make_shared<optical::CoreParams>(std::move(inp));
+}
+
 //---------------------------------------------------------------------------//
 /*!
  * Construct optical tracking offload.
  */
 auto build_optical_offload(inp::Problem const& p,
-                           CoreParams& params,
+                           CoreParams const& params,
                            ImportData const& imported)
 {
-    using optical::MaterialParams;
-
-    CELER_VALIDATE(
-        !imported.optical_materials.empty(),
-        << R"(an optical tracking loop was requested but no optical materials are present)");
     CELER_VALIDATE(p.physics.optical,
                    << "optical physics options are required to construct an "
                       "optical tracking loop");
 
+    OpticalCollector::Input inp;
+    inp.optical_params
+        = build_optical_params(*p.physics.optical, params, imported);
+
+    // Add photon generating processes
+    CELER_ASSERT(p.physics.optical);
     inp::OpticalPhysics const& opt = *p.physics.optical;
-    OpticalCollector::Input oc_inp;
-    oc_inp.material = MaterialParams::from_import(
-        imported, *params.geomaterial(), *params.material());
     if (opt.cherenkov)
     {
-        oc_inp.cherenkov = std::make_shared<CherenkovParams>(*oc_inp.material);
+        inp.cherenkov = std::make_shared<CherenkovParams>(
+            *inp.optical_params->material());
     }
     if (opt.scintillation)
     {
-        oc_inp.scintillation
+        inp.scintillation
             = ScintillationParams::from_import(imported, params.particle());
-        CELER_VALIDATE(oc_inp.scintillation,
+        CELER_VALIDATE(inp.scintillation,
                        << "failed to construct scintillation process");
     }
 
     // Map from optical capacity
     CELER_ASSERT(p.control.optical_capacity);
     inp::OpticalStateCapacity const& cap = *p.control.optical_capacity;
-    auto num_streams = params.max_streams();
-    oc_inp.num_track_slots = ceil_div(cap.tracks, num_streams);
-    oc_inp.buffer_capacity = ceil_div(cap.generators, num_streams);
-    oc_inp.auto_flush = ceil_div(cap.primaries, num_streams);
-    oc_inp.max_step_iters = p.tracking.limits.optical_step_iters;
-
-    // Import models
-    optical::ModelImporter importer{
-        imported, oc_inp.material, params.material()};
-    for (auto const& model : imported.optical_models)
-    {
-        if (auto builder = importer(model.model_class))
-        {
-            oc_inp.model_builders.push_back(*builder);
-        }
-    }
+    auto num_streams = inp.optical_params->max_streams();
+    inp.num_track_slots = ceil_div(cap.tracks, num_streams);
+    inp.buffer_capacity = ceil_div(cap.generators, num_streams);
+    inp.auto_flush = ceil_div(cap.primaries, num_streams);
+    inp.max_step_iters = p.tracking.limits.optical_step_iters;
 
-    CELER_ASSERT(oc_inp);
+    CELER_ENSURE(inp);
```

Comment:
The reason I avoided `inp` here is because it's also a valid namespace; so this is now shadowing `celeritas::inp`.